### DISCUSSION TITLE
Question how to contribut Plugin/ Feature

### DISCUSSION THREAD

**Discussion by jpx40** [rewin123/space_editor#285] [General]
Hey, i wanted to add an text Editor tab with Syntax Highlighting, but i don't really know how to add an Feature. First time using egui. Just Found some egui widgets for a basis.

**Comment by naomijub** [comment]
Do you have a wireframe, picture or drawing so I can understand better the idea?

**Reply by jpx40** [reply]
not yet, but i can make one. My idea was just to have the ability to edit code directly in space_edtior. Maybe with lsp later might be able to copy some of helix code

**Comment by jpx40** [comment]
> Do you have a wireframe, picture or drawing so I can understand better the idea?

https://excalidraw.com/#room=5f6632bb22a3aa6b06e9,Ylrnl_rhtpCBciSjIIABlQ

**Comment by naomijub** [comment]
I see, I would suggest looking into these projects/examples
https://github.com/p4ymak/egui_code_editor
https://github.com/emilk/egui/blob/master/crates/egui_demo_lib/src/demo/code_editor.rs

I do have a plan to be able to read all systems in use and go directly to the in vscode upon click. But that is a feature for the next 2 releases

**Reply by jpx40** [reply]
oh i am an neovim user some integration with terminal or cli stuff would be cool. could you make it so that you can choose the editor. maybe through an cli command that get executed. So the user can use their preferred ide  have space_editor open on the site

**Reply by naomijub** [reply]
I plan to support other integrations, but as I am emacs and vscode user, I'm bit limited there. Once initial integration is done, it should be easy to register new ones.

@rewin123

**Reply by jpx40** [reply]
yeah, if u need help when the first integretton is ready to add intellij or neovim i would be ready. (i am using both so it's in my interest to have them).

**Reply by naomijub** [reply]
I will try to get that included sooner

**Comment by rewin123** [comment]
There are two ways to add a feature to the ui:
- If it is the main functionality of the editor, the code is added directly to crates/space_editor_ui
- If it is an optional functionality, a new crate is created in modules/ folder.

Both ways involve creating a PR in the v0.6 branch

In order to add a new editor panel, we need to use the method

app. editor_tab_by_trait(tab ) 

Which takes a structure that implements the EditorTab trait.

**Reply by jpx40** [reply]
thanks
